web/html/src/components/table/HierarchicalTable.tsx

```diff
@@ -0,0 +1,245 @@
+import * as React from "react";
+import { useState } from "react";
+
+import { SearchField } from "./SearchField";
+import { Table, TableRef } from "./Table";
+
+export type HierarchicalRow = {
+  id: string | number;
+  parentId?: string | number | null;
+  level?: number;
+  expanded?: boolean;
+  isLeaf?: boolean;
+  [key: string]: any;
+};
+
+type HierarchicalTableProps = {
+  /**
+   * Array of data items where each element has a unique id and optional parentId
+   */
+  data: Array<HierarchicalRow>;
+
+  /** Function extracting the unique key of the row from the data object */
+  identifier: (row: HierarchicalRow) => string | number;
+
+  /** The search field to show on the table  */
+  searchField?: React.ReactComponentElement<typeof SearchField>;
+
+  /** Other filter fields */
+  additionalFilters?: Array<React.ReactNode>;
+
+  /** Function to determine which column has expand/collapse controls */
+  expandColumnKey?: string;
+
+  /** Initial expanded state of all nodes (true = all expanded, false = all collapsed) */
+  initiallyExpanded?: boolean;
+
+  /** The column key name of the initial sorted column */
+  initialSortColumnKey?: string;
+
+  /** 1 for ascending, -1 for descending */
+  initialSortDirection?: number;
+
+  /** a function that return a css class for each row */
+  cssClassFunction?: Function;
+
+  /** enables item selection. */
+  selectable?: boolean | ((row: any) => boolean);
+
+  /** the handler to call when the table selection is updated */
+  onSelect?: (items: Array<any>) => void;
+
+  /** the identifiers for selected items */
+  selectedItems?: Array<any>;
+
+  /** Indent size in pixels per level */
+  indentSize?: number;
+
+  /** Children node in the table (Column components) */
+  children: React.ReactNode;
+};
+
+export const HierarchicalTable = React.forwardRef<TableRef, HierarchicalTableProps>((props, ref) => {
+  const {
+    data,
+    identifier,
+    expandColumnKey,
+    initiallyExpanded = false,
+    indentSize = 20,
+    children,
+    ...tableProps
+  } = props;
+
+  // Initialize row expansion state
+  const [expandedRows, setExpandedRows] = useState<Record<string | number, boolean>>(() => {
+    const initialState: Record<string | number, boolean> = {};
+    data.forEach((row) => {
+      initialState[identifier(row)] = initiallyExpanded;
+    });
+    return initialState;
+  });
+
+  // Build the tree structure
+  const buildTreeStructure = (items: Array<HierarchicalRow>) => {
```

Comment:
I don't think I'm understanding the intent here, why do we do all of this on the frontend?